### ISSUE TITLE
ErlangError: :badmatch

### ISSUE THREAD

**Issue by thmzlt** [honeybadger-io/honeybadger-elixir#172]
Here's a screenshot of the Honeybadger stacktrace:

![2018-08-06-121619_1151x912_scrot](https://user-images.githubusercontent.com/7709/43725381-c090b454-9972-11e8-878a-418f6ef9c9a2.png)

Running honeybadger-elixir version 0.10.2 on Elixir 1.6.

**Comment by thmzlt** [comment]
BTW, errors seem to be normally reported. It's just that my Haneybadger project is filled with this error.

**Comment by sorentwo** [comment]
@thmzlt How often is this being reported? Does it seem tied to any particular event (restart, process crash, etc.)?

**Comment by thmzlt** [comment]
> How often is this being reported? 

All the time. In our case it has happened ~3 million times since we started using Honeybadger on our Elixir project two months ago.

> Does it seem tied to any particular event (restart, process crash, etc.)?

It seems to be reported whenever an actual error is reported.

**Comment by minhajuddin** [comment]
@thmzlt we'll check this and get back to you

**Comment by sorentwo** [comment]
@minhajuddin I'm on it 👍

**Comment by thmzlt** [comment]
Thanks a lot for looking into it!

**Comment by ericlathrop** [comment]
I'm getting this same error after upgrading honeybadger from `0.10.0` to `0.10.3`, Elixir from `1.6.4` to `1.7.3`, and Erlang from `20` to `21`. The stack traces is the same, but the line numbers are slightly different. 
![screenshot from 2018-09-27 13-43-59](https://user-images.githubusercontent.com/530891/46164325-b3d9ad80-c25b-11e8-80a0-0a5c14835aa1.png)

**Comment by sorentwo** [comment]
I wonder if this is a slightly different error introduced by the automatic logger metadata extraction.

@ericlathrop Is this happening often or in a reproducible way?

**Comment by minhajuddin** [comment]
@sorentwo `0.10.3` doesn't seem to have the logger metadata changes.

**Comment by sorentwo** [comment]
tl;dr — This is a separate issue, but something that I'll be investigating.

I believe what we're seeing dissociated errors that are grouped together by an incorrect stack trace. For example, here is a connection timeout error that was just reported:

![screenshot_2018-10-12 exception and uptime monitoring 360-degree coverage of errors outages and service degradation](https://user-images.githubusercontent.com/270831/46889812-744dbc80-ce2a-11e8-9245-ff3bc801a103.png)

The error report didn't have an associated stacktrace, so the honeybadger client attempted to use `Process.info/1` to grab the most recent stacktrace. That stacktrace doesn't include any actual error information, so the trace is fairly useless.

P.S. — Firefox's image capture stuck the header in the middle of the image, ignore that.

**Comment by axelson** [comment]
Is the/a problem that in the BEAM a stracktrace is generated only when an exception occurs and if you call `System.stacktrace()` or (probably) `Process.info(self(), :current_stacktrace)` you will get an incorrect stacktrace?

So I think that the current stacktrace fetching code may be flawed (although I may be totally off base here):
https://github.com/honeybadger-io/honeybadger-elixir/blob/817d5feb0a98c5ed23bb758e2ffa9671ac4f9aab/lib/honeybadger.ex#L316

**Comment by sorentwo** [comment]
@axelson Yes, that is the culprit. We aren't supposed to hit that clause during regular operation, it is a last resort that is supposed to guarantee _some_ stacktrace is reported.

The situation we seem to be running into is when errors are reported through the error logger. If the logged error doesn't include the original stacktrace then we don't have any way to grab it. At that point we aren't inside of a `try/catch` and `System.stacktrace()` isn't guaranteed to work.

I'm fairly sure that integration is in `lib/honeybadger/logger.ex:48` is the problem. It isn't handling message shapes from other errors like Tasks, gen_event, gen_server, exits, etc.

**Comment by tielur** [comment]
I recently started getting this exact same error on two different elixir applications:
elixir 1.6.5
erlang 20.3
honeybadger 0.10.3

App 1:
![image](https://user-images.githubusercontent.com/977174/55050334-5c9b7680-500d-11e9-8112-ff3e40969178.png)
App 2:
![image](https://user-images.githubusercontent.com/977174/55050347-6b822900-500d-11e9-86dc-87a1e666de2b.png)

**Comment by tielur** [comment]
@sorentwo Any idea why this would still be happening?

**Comment by sorentwo** [comment]
@tielur The fix for that error is in Honeybadger v0.11.0, but it requires Elixir 1.7+ and Erlang 21+. The fix involved an overhaul to how we integrated with logger, which prompted a move to 1.7+.

See the [CHANGELOG](https://github.com/honeybadger-io/honeybadger-elixir/blob/master/CHANGELOG.md#v0110---2019-02-28) and the [Pull Request](https://github.com/honeybadger-io/honeybadger-elixir/pull/204) for more details.

**Comment by tielur** [comment]
@sorentwo thank you!

For what it's worth, I think my error was related to `tzdata` updating and the latest 2019a time zone database update had a typo in it(https://github.com/lau/tzdata/issues/9#issuecomment-476854925)

So I think tzdata was throwing an error, honeybadger had a mismatch but in the grand scheme of things my application kept on running. ❤️ 

I guess it's time to upgrade elixir and erlang! Thanks again!
